lib/Logger/LogContext.h

```diff
@@ -0,0 +1,214 @@
+    }
+    void visit(std::string_view const& key, std::int64_t value) const override {
+      self().visit(key, value);
+    }
+    void visit(std::string_view const& key, std::uint64_t value) const override {
+      self().visit(key, value);
+    }
+
+   private:
+    Derived const& self() const noexcept {
+      return static_cast<Derived const&>(*this);
+    }
+  };
+
+  template <class Overloads>
+  struct OverloadVisitor : GenericVisitor<OverloadVisitor<Overloads>>, Overloads {
+    explicit OverloadVisitor(Overloads overloads) : GenericVisitor<OverloadVisitor>(), Overloads(std::move(overloads)) {}
+    template <class T>
+    void visit(std::string_view const& key, T&& value) const {
+      this->operator()(key, std::forward<T>(value));
+    }
+  };
+
+  struct Values {
+    virtual ~Values() = default;
+    virtual void visit(Visitor const&) = 0;
+  };
+
+  /// @brief helper class to create value tuples with the following syntax:
+  ///   LogContext::makeValue().with<Key1>(value1).with<Key2>(value2)
+  /// The keys must be compile time constant char pointers; the values can
+  /// be strings or numbers.
+  ///   TODO: support for arbitrary types that support operator<<
+  /// As shown, the inital ValueBuilder instance can be created via LogContext::makeValue().
+  /// The `ValueBuilder` can be passed directly to `ScopedValue`, or you
+  /// can call `share()` to create a reusable `std::shared_ptr<Values>`
+  /// which can be stored in a member and used for different `ScopedValues`.
+  template <class Values = std::tuple<>, const char... Keys[]>
+  struct ValueBuilder {
+    template <const char Key[], class Value>
+    auto with(Value v) && {
+      auto vals = std::tuple_cat(std::move(_vals), std::make_tuple(std::forward<Value>(v)));
+      return ValueBuilder<decltype(vals), Keys..., Key>(std::move(vals));
+    }
+    std::shared_ptr<Values> share() && {
+      return std::make_shared<ValuesImpl<Values, Keys...>>(std::move(_vals));
+    }
+   private:
+    friend class LogContext;
+    ValueBuilder(Values vals) : _vals(std::move(vals)) {}
+    Values _vals;
+  };
+
+  static ValueBuilder<> makeValue() { return ValueBuilder<std::tuple<>>({}); }
+
+  /// @brief adds the provided value(s) to the LogContext for the current scope,
+  /// i.e., upon destruction the value(s) are removed from the current LogContext.
+  struct ScopedValue {
+    template <class Values, const char... Keys[]>
+    ScopedValue(ValueBuilder<Values, Keys...>&& v) {
+      appendEntry(std::make_shared<EntryImpl<ValuesImpl<Values, Keys...>>>(std::move(v._vals)));
+    }
+
+    ScopedValue(std::shared_ptr<Values>&& v) {
+      appendEntry(std::make_shared<EntryImpl<std::shared_ptr<Values>>>(std::move(v)));
+    }
+
+    ~ScopedValue();
+  private:
+    void appendEntry(std::shared_ptr<Entry> e) {
+      _oldTail = e.get();
+      LogContext::current().pushEntry(std::move(e));
+    }
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+    Entry* _oldTail;
```

Comment:
Looks safer to me if we initialize it to a defined value rather than leaving it undefined.
```suggestion
    Entry* _oldTail = nullptr;
```